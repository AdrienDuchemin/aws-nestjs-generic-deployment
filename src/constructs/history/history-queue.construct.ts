import { Queue, QueueProps } from '@aws-cdk/aws-sqs'
import { Construct } from '@aws-cdk/core'
import { History } from './history.construct'

export type IQueueProps = Omit<QueueProps, 'fifo'>

export interface IHistoryQueueProps {
  historyQueueProps?: IQueueProps
}

export class HistoryQueue extends Queue {
  constructor(scope: Construct, id: string, props: IHistoryQueueProps) {
    const { historyQueueProps } = props
    const { queueName = `${id}-history-queue-fifo` } = historyQueueProps ? historyQueueProps : {}

    super(scope, id, { 
      fifo: true,
      ...historyQueueProps,
      queueName
    })

    new History(this, id, {
      queue: this
    })
  }
}