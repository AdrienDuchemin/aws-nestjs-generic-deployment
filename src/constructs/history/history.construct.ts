import { join } from 'path'
import { Table, AttributeType, TableProps } from '@aws-cdk/aws-dynamodb'
import { Code } from '@aws-cdk/aws-lambda'
import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources'
import { Queue } from '@aws-cdk/aws-sqs'
import { Construct } from '@aws-cdk/core'
import { Lambda } from '../lambda.construct'

export interface HistoryProps {
  queue: Queue,
  tableProps?: TableProps
}

export class History extends Construct {
  constructor(scope: Construct, id: string, props: HistoryProps) {
    const { queue, tableProps } = props
    const { tableName = `${id}HistoryTable`, partitionKey = { name: 'id', type: AttributeType.STRING } } = tableProps ? tableProps : {}

    super(scope, id)

    const lambda = new Lambda(this, `${id}History`, {
      lambdaProps: {
        code: Code.fromAsset(join(__dirname, './lambda')),
        handler: 'index.handler',
      }
    })
    
    lambda.addEventSource(new SqsEventSource(queue))
    
    const table = new Table(this, `${id}HistoryTable`, {
      ...tableProps,
      tableName,
      partitionKey
    })
    table.grantReadWriteData(lambda)

    lambda.addEnvironment('TABLE_NAME', table.tableName)
  }
}