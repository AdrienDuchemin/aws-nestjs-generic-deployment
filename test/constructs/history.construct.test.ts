import { expect as expectCDK, SynthUtils, countResources, haveResource } from '@aws-cdk/assert'
import { Queue } from '@aws-cdk/aws-sqs'
import { Stack } from '@aws-cdk/core'
import { History } from '../../src/constructs/history/history.construct'

describe('History', () => {
  let stack: Stack

  beforeAll(()=> {
    stack = new Stack()
    new History(stack, 'LambdaHistory', {
      queue: new Queue(stack, 'HistoryQueue')
    })
  })

  it('should create one Lambda', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1))
    expectCDK(stack).to(haveResource("AWS::Lambda::Function",{
      Handler: 'index.handler'
    }))
  })

  it('should create one Table', () => {
    expectCDK(stack).to(countResources("AWS::DynamoDB::Table", 1))
    expectCDK(stack).to(haveResource("AWS::DynamoDB::Table",{
      KeySchema: [{ 
        AttributeName: "id",
        KeyType: "HASH",
      }],
    }))
  })

  it('should not have changed', () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})