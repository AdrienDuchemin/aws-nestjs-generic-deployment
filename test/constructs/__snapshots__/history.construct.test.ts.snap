// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History should not have changed 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05ArtifactHash6241674D": Object {
      "Description": "Artifact hash for asset \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3Bucket3765D8AA": Object {
      "Description": "S3 bucket for asset \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
    "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835": Object {
      "Description": "S3 key for asset version \\"6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HistoryQueue16C3A71A": Object {
      "Type": "AWS::SQS::Queue",
    },
    "LambdaHistoryLambdaHistoryHistoryLambdaFF6E0B90": Object {
      "DependsOn": Array [
        "LambdaHistoryLambdaHistoryHistoryLambdaServiceRoleDefaultPolicy0084711D",
        "LambdaHistoryLambdaHistoryHistoryLambdaServiceRole92BB1E40",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3Bucket3765D8AA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6e299eb822cce449ea66aea74a07269cec00aa22054e7820f61d319cb0031e05S3VersionKey59FE3835",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "LambdaHistoryLambdaHistoryHistoryTable8A26FC1C",
            },
          },
        },
        "FunctionName": "LambdaHistoryHistoryLambda",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaHistoryLambdaHistoryHistoryLambdaServiceRole92BB1E40",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaHistoryLambdaHistoryHistoryLambdaServiceRole92BB1E40": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaHistoryLambdaHistoryHistoryLambdaServiceRoleDefaultPolicy0084711D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "HistoryQueue16C3A71A",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "LambdaHistoryLambdaHistoryHistoryTable8A26FC1C",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaHistoryLambdaHistoryHistoryLambdaServiceRoleDefaultPolicy0084711D",
        "Roles": Array [
          Object {
            "Ref": "LambdaHistoryLambdaHistoryHistoryLambdaServiceRole92BB1E40",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaHistoryLambdaHistoryHistoryLambdaSqsEventSourceHistoryQueueD3A6A211": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "HistoryQueue16C3A71A",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "LambdaHistoryLambdaHistoryHistoryLambdaFF6E0B90",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "LambdaHistoryLambdaHistoryHistoryTable8A26FC1C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "LambdaHistoryHistoryTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
